staging/src/k8s.io/client-go/tools/cache/delta_fifo_test.go

```diff
@@ -121,6 +121,35 @@ func TestDeltaFIFO_replaceWithDeleteDeltaIn(t *testing.T) {
 	}
 }
 
+// TestDeltaFIFO_replaceWithAddDeltaIn tests that a `Deleted` delta for an
+// object `O` with ID `X` is added when .Replace is called and `O` is not
+// among the replacement objects, and not present in the backing store,
+// but some delta for the object is present in the DeltaFIFO.
+// Not adding a `Deleted` delta would cause the delete to be missed,
+// and the informer won't get rid of it before the next .Replace/relist.
+func TestDeltaFIFO_replaceWithAddDeltaIn(t *testing.T) {
+	newObj := mkFifoObj("foo", 2)
+
+	f := NewDeltaFIFOWithOptions(DeltaFIFOOptions{
+		KeyFunction: testFifoObjectKeyFunc,
+		KnownObjects: literalListerGetter(func() []testFifoObject {
+			return []testFifoObject{}
+		}),
+	})
+
+	f.Add(newObj)
```

Comment:
You need to test all possible deltas -- add, update, sync (?), and delete. I think this code does the wrong thing if there's a delete already there?

Reply:
> You need to test all possible deltas -- add, update, sync (?), and delete.

Yeah, smart! Will add another set of tests.

>  I think this code does the wrong thing if there's a delete already there?

If there is a delete there already, and the object should be deleted, it should be on the tail of the deltas. And if that is the case, the new "delete" event will be a noop since its deduped here I [believe](https://github.com/kubernetes/kubernetes/blob/v1.26.0/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L396-L405).

Reply:
Ahh, yee, I think you are correct. If the delete event is a normal delete event its fine, but if the object is a `
DeletedFinalStateUnknown` we have to extract the actual object. Or can we just `continue` the loop? I think for the end users both should be the same.

Reply:
OOh, well, that uncover the same "issue" for the case where `f.knownObjects == nil` as well. I'll write a test to verify the behavior.